// @rainbow-me/rainbowkit@2.1.7/dist/wallets/walletConnectors/chunk-PHF4VWKP.js
"use client";
import {
  getInjectedConnector,
  hasInjectedProvider
} from "./chunk-MBBGZGXF.js";

// src/wallets/walletConnectors/safeheronWallet/safeheronWallet.ts
var safeheronWallet = () => ({
  id: "safeheron",
  name: "Safeheron",
  installed: hasInjectedProvider({
    namespace: "safeheron",
    flag: "isSafeheron"
  }),
  iconUrl: async () => (await import("./safeheronWallet-6JG77AKC.js")).default,
  iconBackground: "#fff",
  downloadUrls: {
    chrome: "https://chrome.google.com/webstore/detail/safeheron/aiaghdjafpiofpainifbgfgjfpclngoh",
    browserExtension: "https://www.safeheron.com/"
  },
  extension: {
    instructions: {
      learnMoreUrl: "https://www.safeheron.com/",
      steps: [
        {
          description: "wallet_connectors.safeheron.extension.step1.description",
          step: "install",
          title: "wallet_connectors.safeheron.extension.step1.title"
        },
        {
          description: "wallet_connectors.safeheron.extension.step2.description",
          step: "create",
          title: "wallet_connectors.safeheron.extension.step2.title"
        },
        {
          description: "wallet_connectors.safeheron.extension.step3.description",
          step: "refresh",
          title: "wallet_connectors.safeheron.extension.step3.title"
        }
      ]
    }
  },
  createConnector: getInjectedConnector({
    namespace: "safeheron",
    flag: "isSafeheron"
  })
});

export {
  safeheronWallet
};
