// @rainbow-me/rainbowkit@2.1.7/dist/wallets/walletConnectors/chunk-DCT72KYP.js
"use client";
import {
  isAndroid
} from "./chunk-RETKWSKD.js";
import {
  getWalletConnectConnector
} from "./chunk-75A5A7MH.js";

// src/wallets/walletConnectors/argentWallet/argentWallet.ts
var argentWallet = ({
  projectId,
  walletConnectParameters
}) => ({
  id: "argent",
  name: "Argent",
  iconUrl: async () => (await import("./argentWallet-M6SBG4CU.js")).default,
  iconBackground: "#fff",
  downloadUrls: {
    android: "https://play.google.com/store/apps/details?id=im.argent.contractwalletclient",
    ios: "https://apps.apple.com/us/app/argent/id1358741926",
    mobile: "https://argent.xyz/download-argent",
    qrCode: "https://argent.link/app"
  },
  mobile: {
    getUri: (uri) => {
      return isAndroid() ? uri : `argent://app/wc?uri=${encodeURIComponent(uri)}`;
    }
  },
  qrCode: {
    getUri: (uri) => uri,
    instructions: {
      learnMoreUrl: "https://argent.xyz/learn/what-is-a-crypto-wallet/",
      steps: [
        {
          description: "wallet_connectors.argent.qr_code.step1.description",
          step: "install",
          title: "wallet_connectors.argent.qr_code.step1.title"
        },
        {
          description: "wallet_connectors.argent.qr_code.step2.description",
          step: "create",
          title: "wallet_connectors.argent.qr_code.step2.title"
        },
        {
          description: "wallet_connectors.argent.qr_code.step3.description",
          step: "scan",
          title: "wallet_connectors.argent.qr_code.step3.title"
        }
      ]
    }
  },
  createConnector: getWalletConnectConnector({
    projectId,
    walletConnectParameters
  })
});

export {
  argentWallet
};
