// @rainbow-me/rainbowkit@2.1.7/dist/wallets/walletConnectors/chunk-DVXPOWEC.js
"use client";
import {
  getInjectedConnector,
  hasInjectedProvider
} from "./chunk-MBBGZGXF.js";

// src/wallets/walletConnectors/desigWallet/desigWallet.ts
var desigWallet = () => {
  return {
    id: "desig",
    name: "Desig Wallet",
    iconUrl: async () => (await import("./desigWallet-WJMSF3PS.js")).default,
    iconBackground: "#ffffff",
    installed: hasInjectedProvider({ namespace: "desig.ethereum" }),
    downloadUrls: {
      android: "https://play.google.com/store/apps/details?id=io.desig.app",
      ios: "https://apps.apple.com/app/desig-wallet/id6450106028",
      qrCode: "https://desig.io",
      mobile: "https://desig.io",
      browserExtension: "https://chrome.google.com/webstore/detail/desig-wallet/panpgppehdchfphcigocleabcmcgfoca"
    },
    extension: {
      instructions: {
        steps: [
          {
            description: "wallet_connectors.desig.extension.step1.description",
            step: "install",
            title: "wallet_connectors.desig.extension.step1.title"
          },
          {
            description: "wallet_connectors.desig.extension.step2.description",
            step: "create",
            title: "wallet_connectors.desig.extension.step2.title"
          },
          {
            description: "wallet_connectors.desig.extension.step3.description",
            step: "refresh",
            title: "wallet_connectors.desig.extension.step3.title"
          }
        ],
        learnMoreUrl: "https://desig.io"
      }
    },
    createConnector: getInjectedConnector({
      namespace: "desig.ethereum"
    })
  };
};

export {
  desigWallet
};
