// @web3modal/base@5.1.10/dist/esm/adapters/solana/web3js/utils/defaultConfig.js
import '@web3modal/polyfills';
export function defaultSolanaConfig(options) {
    const { enableInjected = true, metadata } = options;
    let injectedProvider = undefined;
    const providers = { metadata, auth: options.auth };
    function getInjectedProvider() {
        if (injectedProvider) {
            return injectedProvider;
        }
        if (typeof window === 'undefined') {
            return undefined;
        }
        if (!window.solana) {
            return undefined;
        }
        injectedProvider = window.solana;
        return injectedProvider;
    }
    if (enableInjected) {
        providers.injected = getInjectedProvider();
    }
    const defaultAuth = {
        email: true,
        showWallets: true,
        walletFeatures: true,
        socials: [
            'google',
            'x',
            'discord',
            'farcaster',
            'github',
            'apple',
            'facebook'
        ]
    };
    providers.auth = {
        ...defaultAuth,
        ...options.auth
    };
    return providers;
}
